File: Decembrie-2023/Meditație 9.12.2023/QuickSort.cpp
#include<iostream>
#define FIN "algsort.in"
#define FOUT "algsort.out"
using namespace std;
int v[500000], n ;
void read()
{
	freopen(FIN, "r", stdin);
	scanf("%d", &n);
	for(int i=0; i<n; i++)
		scanf("%d", &v[i]);
}
void quick_sort(int low, int high)
{
	int pivot = v[(low+high)>>1],
		i=low, j=high;
	while(i<=j){
		while( v[i]< pivot ) 
			i++;
		while( v[j]> pivot )
			j--;
		if(i<=j) {
			int temp = v[i];
			v[i] = v[j];
			v[j] = temp;
			i++; j--;
		}
 	}
// 				i>>>>	high
// low 	j<<<<
	if( low < j ) quick_sort(low, j);
	if( i < high ) quick_sort(i, high );
}

void write()
{
	freopen(FOUT, "w", stdout);
	for(int i=0; i<n; i++)
		printf("%d ", v[i] );
}
int main()
{
	read();
	quick_sort(0, n-1);
	write();
}

File: Decembrie-2023/Meditație 13.12.2023/QuickSort.java
public class QuickSort {
	public static void main(String[] args) {
		int[] arr = {9, 8, 7, 6, 5, 4, 3, 2, 2, -3, 0};
	System.out.println("Vectorul initial:");
	display(arr);
	quicksort(0, arr.length-1, arr);
	System.out.println("Vectorul sortat:");
	display(arr);
	}

	public static void quicksort(int low,int high,int arr[])
{
			int i= low,
			j = high,
			pivot = arr[(low+high)>>1];
		while(i<=j){
			while(arr[i]<pivot) 
				i++;
			while(arr[j]>pivot)
				j--;
			if(i<=j){
				int temp = arr[i];
				arr[i] = arr[j];
				arr[j] = temp;
				i++; j--;
			}
		}
		if(low<j) quicksort(low, j, arr);
		if(i<high) quicksort(i,high, arr);
	}
	public static void display(int[] arr){
		for(int value: arr)
			System.out.print(value+ " ");
		System.out.println();
	
	}
}